numpy/ma/core.py

```diff
@@ -6300,6 +6306,12 @@ def __deepcopy__(self, memo=None):
         memo[id(self)] = copied
         for (k, v) in self.__dict__.items():
             copied.__dict__[k] = deepcopy(v, memo)
+        # as clearly documented for np.copy(), you need to use
+        # deepcopy() directly for arrays of object type that may
+        # contain compound types--you cannot depend on normal
+        # copy semantics to do the right thing here
+        if self.dtype.hasobject:
+            copied = deepcopy(self._data)
```

Comment:
```suggestion
            copied._data = deepcopy(copied._data)
```
It looks like you are returning the data array only otherwise :).

Reply:
Looks like `_data` doesn't have a "setter," but maybe you can live with my revision.

Since the test suite was blind to it I also beefed up one of the tests to be sensitive to it and check the `_mask` deepcopy too.

Reply:
Oh, interesting, I thought it is a private attribute and fine to just replace it entirely.  This seems good, though.  Except that it should be `...` to support 0-D arrays.

(Might just do that and merge some time, feels a bit like a 95% fix, but this is MA, 95% fixes are probably about as good as it gets :))